Use Vazir font with a node module instead of CDN

NeOMakinG opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Using a CDN exposes users to a risk of data abuse and privacy abuse, so we swap to a local font using npm
| Type?             | improvement
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #27541.
| How to test?      | RTL language font should be loaded normally (tested it, works fine), also when you run `npm run build` on default, new-theme and classic/_dev folder, it should generate a thirdPartyNotice.json file in the asset folder, containing every notices
| Possible impacts? | RTL font


<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/27611)
<!-- Reviewable:end -->

Commit:
Use Vazir font with a node module instead of CDN

Commit:
Fix scss lint

Commit:
Run CSFixer

Commit:
Add license notice generator

Commit:
Fix lin

Commit:
Run only on build

NeOMakinG commented:
Can be QA by a dev I think so!

Commit:
Run scss-lint

Commit:
Run scss-lint

Commit:
Fix js

matks reviewed (dismissed):
I remember we discussed using a CDN, yes or no, for another topic.

Maybe we should add a mention in devdocs FAQ about this?

NeOMakinG commented:
> I remember we discussed using a CDN, yes or no, for another topic.
> 
> Maybe we should add a mention in devdocs FAQ about this?

could probably be useful yes

mparvazi commented:
Great improvement :100: 

Some suggestions:

1. Just need to use `woff` and `woff2` for font-face. Why you should be responsible for customers who are stuck in the past? (IE or Opera Mini for BO :neutral_face: ). Even Microsoft has been planned to get rid of these type of users.
```
@font-face {
  font-family: Vazir;
  font-style: normal;
  font-display: swap;
  font-weight: 300;
  src: url("~vazir-font/dist/Vazir-Light.woff2") format("woff2"), url("~vazir-font/dist/Vazir-Light.woff") format("woff");
}
@font-face {
  font-family: Vazir;
  font-style: normal;
  font-display: swap;
  font-weight: 400;
  src: url("~vazir-font/dist/Vazir-Regular.woff2") format("woff2"), url("~vazir-font/dist/Vazir-Regular.woff") format("woff");
}
@font-face {
  font-family: Vazir;
  font-style: normal;
  font-display: swap;
  font-weight: 700;
  src: url("~vazir-font/dist/Vazir-Bold.woff2") format("woff2"), url("~vazir-font/dist/Vazir-Bold.woff") format("woff");
}
```
_Fonts will be reduced from 15 to 6 in public folder._

2. The `rtl_rtl.css` will be generated by CSSJanus in public folder, but only `rtl.css` will be used, so need to prevent transforming `rtl.scss` by CSSJanus. Because in future you may put overrides of RTL inside `rtl.scss` (plan to stop using theme.rtlfix).
`src/Core/Localization/RTL/StylesheetGenerator.php`
![proccess](https://user-images.githubusercontent.com/85633460/148949309-0514f991-0683-4565-a552-c2456b1a229f.png)

3. We don't have any call for `rtl.css` in Login page, so the `font-face` doesn't use in page. I think need to add extra call for Login page in `AdminController.php` but I'm not sure where!
![Screenshot from 2022-02-26 15-55-55](https://user-images.githubusercontent.com/85633460/155843110-6da4cbcc-17a4-48a2-9f21-67ee2c63d217.png)

mparvazi commented:
Great improvement :100: 

Some suggestions:

1- Just need to use `woff` and `woff2` for font-face (to me `woff2` is enough!)
Why you should be responsible for users who are stuck in the past? (`eot` only uses by IE, `ttf` only needed for very old browsers). Even Microsoft has been planned to get rid of these users.
IMO a browser that doesn't support `woff2`, also it doesn't support most of components in BO, so why we should always care about font-face support?
```
@font-face {
  font-family: Vazir;
  font-style: normal;
  font-display: swap;
  font-weight: 300;
  src: url("~vazir-font/dist/Vazir-Light.woff2") format("woff2"), url("~vazir-font/dist/Vazir-Light.woff") format("woff");
}

@font-face {
  font-family: Vazir;
  font-style: normal;
  font-display: swap;
  font-weight: 400;
  src: url("~vazir-font/dist/Vazir-Regular.woff2") format("woff2"), url("~vazir-font/dist/Vazir-Regular.woff") format("woff");
}

@font-face {
  font-family: Vazir;
  font-style: normal;
  font-display: swap;
  font-weight: 700;
  src: url("~vazir-font/dist/Vazir-Bold.woff2") format("woff2"), url("~vazir-font/dist/Vazir-Bold.woff") format("woff");
}
```
_The number of fonts will be reduced from 15 to 6 in the public folder._ 

2- The `rtl_rtl.css` will be generated by CSSJanus in public folder, but only `rtl.css` will be used, so need to prevent transforming `rtl.css` by CSSJanus (PHP plugin that use by `StylesheetGenerator.php`). Because in future you may put overrides of RTL inside `rtl.scss` (plan to stop using theme.rtlfix).
`src/Core/Localization/RTL/StylesheetGenerator.php`
![proccess](https://user-images.githubusercontent.com/85633460/148949309-0514f991-0683-4565-a552-c2456b1a229f.png)

3- We don't have any call for `rtl.css` in Login page:
`controllers/admin/AdminLoginController.php`
![Screenshot from 2022-02-27 11-41-24](https://user-images.githubusercontent.com/85633460/155874311-93ee99a5-3be6-4aff-a0a8-2cb0d92c236f.png)

4- The license text for [vazir-font](https://github.com/rastikerdar/vazir-font) is null. I think it's related to missing "LICENSE" file in project's root folder.
![Screenshot from 2022-02-26 16-21-05](https://user-images.githubusercontent.com/85633460/155844019-75549ab0-cbae-4c1b-980f-b561bed83ea1.png)
Can use `replenishDefaultLicenseTexts` option (when this is enabled, default license texts are taken from `spdx.org` for packages where no license text was found):
![Screenshot from 2022-02-27 11-43-34](https://user-images.githubusercontent.com/85633460/155874384-b560aefc-a0b4-4dc1-a34b-ac2d74db1915.png)

5- After webpack build for default theme, there is an empty `rtl.bundle.js` file, so we can use [webpack-remove-empty-scripts](https://github.com/webdiscus/webpack-remove-empty-scripts).

Commit:
Remove other fonts than woff2 and add rtl css to login

Commit:
Add default license text

Commit:
Lint js

Commit:
Remove empty scripts with webpack

NeOMakinG commented:
@mparvazi Every points are fixed, except the rtl.css with CSSJanus, actually we can't do it, it's not a problem for the moment but we need to take care about the fact that no one add something else than fonts inside rtl.scss

mparvazi commented:
> @mparvazi Every points are fixed, except the rtl.css with CSSJanus, actually we can't do it, it's not a problem for the moment but we need to take care about the fact that no one add something else than fonts inside rtl.scss

I'm the one who want to add override rules for RTL inside `rtl.scss`  :smile: 
Then there is no need for `theme.rtlfix` for BO theme.

mparvazi commented:
@NeOMakinG 
I tested and works fine.

matks commented:
@Robin-Fischer-PS @NeOMakinG I suggest a dev does the QA?

Robin-Fischer-PS commented:
@matks @NeOMakinG I need some more info : Is there a high risk of regressions on RTL display ? If so, I would like to have a regression test made by one of the QA girls.

NeOMakinG commented:
> @matks @NeOMakinG I need some more info: Is there a high risk of regressions on RTL display? If so, I would like to have a regression test made by one of the QA girls.

- RTL won't work fine anymore on IE11 (which is not supported anymore in june) and some very old browsers with low usage
- No high risk of regressions on other browsers

Robin-Fischer-PS commented:
OK for QA by dev then ;)

Thanks for the quick answer @NeOMakinG :) (and sorry for my late answer, I forgot to respond yesterday :man_facepalming: )

mparvazi commented:
@NeOMakinG 

The `Vazir` font renamed to `Vazirmatn` and its structure has been changed (this font going to be add to the Google fonts soon, so its structure changed based on Google standard).

https://github.com/rastikerdar/vazirmatn

```
@font-face {
  font-family: Vazirmatn;
  src: url('~vazirmatn/fonts/webfonts/Vazirmatn-Light.woff2') format('woff2');
  font-weight: 300;
  font-style: normal;
  font-display: swap;
}

@font-face {
  font-family: Vazirmatn;
  src: url('~vazirmatn/fonts/webfonts/Vazirmatn-Regular.woff2') format('woff2');
  font-weight: 400;
  font-style: normal;
  font-display: swap;
}

@font-face {
  font-family: Vazirmatn;
  src: url('~vazirmatn/fonts/webfonts/Vazirmatn-Bold.woff2') format('woff2');
  font-weight: 700;
  font-style: normal;
  font-display: swap;
}
```

NeOMakinG commented:
converted to draft as @mparvazi warned about something, let's wait

mparvazi commented:
@NeOMakinG 

Fonts added to Google fonts:
`https://github.com/google/fonts/pull/4352`

So it's ready to use.

Progi1984 reviewed (changes requested):
Needs a rebase and Shouldn't be a draft.

mparvazi commented:
@NeOMakinG 

Ready to use:
https://github.com/rastikerdar/vazirmatn

Commit:
fix: use new vazir font on BO

Commit:
fix: remove classic things

NeOMakinG commented:
> Needs a rebase and Shouldn't be a draft.

The draft mode was needed because we were blocking the PR, vazir has been updated to a new package.

I removed classic stuff and changed the vazir package, it's ready for review now @mparvazi and @PrestaShop/prestashop-maintainers

Commit:
fix: problem on new theme

Commit:
fix: lint

mparvazi commented on admin-dev/themes/new-theme/package.json:
```diff
@@ -42,15 +42,16 @@
     "terser-webpack-plugin": "^5.2.5",
     "typeahead.js": "^0.11.1",
     "url-polyfill": "^1.1.7",
+    "vazirmatn": "^32.0.0",
```
```suggestion
    "vazirmatn": "^32.1.0",
```

mparvazi commented on admin-dev/themes/default/package.json:
```diff
@@ -46,11 +46,14 @@
     "open-sans-fonts": "^1.6.2",
-    "perfect-scrollbar": "^1.4.0"
+    "perfect-scrollbar": "^1.4.0",
+    "vazirmatn": "^32.0.0"
```
```suggestion
    "vazirmatn": "^32.1.0"
```

mparvazi commented on admin-dev/themes/default/webpack.config.js:
```diff
@@ -118,6 +122,13 @@ module.exports = (env, argv) => {
+      new LicensePlugin({ 
+        outputFilename: 'thirdPartyNotice.json',
+        licenseOverrides: {
+          'vazirmatn@32.0.0': 'OFL-1.1'
```
```suggestion
          'vazirmatn@32.1.0': 'OFL-1.1'
```

mparvazi commented on admin-dev/themes/new-theme/.webpack/common.js:
```diff
@@ -403,5 +405,13 @@ module.exports = {
+    new LicensePlugin({
+      outputFilename: 'thirdPartyNotice.json',
+      licenseOverrides: {
+        'vazirmatn@32.0.0': 'OFL-1.1',
```
```suggestion
        'vazirmatn@32.1.0': 'OFL-1.1',
```

mparvazi commented:
There is a problem in v32.0.0 about kerning in numbers.
That issue is related to different displaying in windows 10 and Linux.

Fixed in v32.1

Commit:
fix: update vazir to 32.1.0

Commit:
fix: send lock files

Progi1984 commented on admin-dev/themes/new-theme/scss/rtl.scss:
```diff
@@ -0,0 +1,21 @@
+  font-weight: 400;
+  src: url("~vazirmatn/fonts/webfonts/Vazirmatn-Regular.woff2") format("woff2");
+}
+
+
```
```suggestion

```

Commit:
Unneeded space

Co-authored-by: Progi1984 <progi1984@gmail.com>

matks commented:
I checked out the branch and tested. As far as I see it works as expected.
